edx_rest_framework_extensions/tests/test_utils.py

```diff
@@ -38,10 +71,81 @@ def setUp(self):
+        """
+        Validates that a valid token is properly decoded given a list of multiple valid issuers
+        """
+
+        # Verify that each valid issuer is properly matched against the valid issuers list
+        # and used to decode the token that was generated using said valid issuer data
+        self.payload['iss'] = jwt_issuer['ISSUER']
+        token = generate_jwt_token(self.payload, jwt_issuer['SECRET_KEY'])
+        self.assertEqual(utils.jwt_decode_handler(token), self.payload)
 
-    def test_decode_error(self):
+    @override_settings(JWT_AUTH=JWT_AUTH_VALID_ISSUERS)
+    def test_decode_raises_invalid_token_error_invalid_token(self):
+        """
+        Validates that an invalid token raises the correct exception, and writes the correct
+        exception information to the application's log file.
+        """
+
+        # The handler will iterate over all of the issuers, attempting to decode the token,
+        # but since it's an invalid token to start, the workflow will ultimately fail and
+        # should raise an InvalidTokenError, which our test will verify.
         with mock.patch('edx_rest_framework_extensions.utils.logger') as patched_log:
             with self.assertRaises(jwt.InvalidTokenError):
-                jwt_decode_handler('not.a.valid.jwt')
-            patched_log.exception.assert_called_once_with('JWT decode failed!')
+                utils.jwt_decode_handler('invalid.token')
+            # Verify that the proper entries were written to the log
+            msg = "Token decode failed for issuer 'valid-issuer'"
+            patched_log.info.assert_any_call(msg)
+            msg = "Token decode failed for issuer 'another-valid-issuer'"
+            patched_log.info.assert_any_call(msg)
+            msg = "All combinations of JWT issuers and secret keys failed to validate the token."
+            patched_log.error.assert_any_call(msg)
+
+    @ddt.data(*JWT_ISSUERS)
+    @override_settings(JWT_AUTH=JWT_AUTH_VALID_ISSUERS)
+    def test_decode_raises_improperly_configured_exception_no_jwt_issuers_defined(self, jwt_issuer):
+        """
+        Validates that an ImproperlyConfigured exception is raised when no JWT_ISSUERS are set
+        """
+        with mock.patch("edx_rest_framework_extensions.utils.get_jwt_issuers") as mock_get_jwt_issuers:
+            with self.assertRaises(ImproperlyConfigured):
+                mock_get_jwt_issuers.return_value = None
+                self.payload['iss'] = jwt_issuer['ISSUER']
+                signing_key = jwt_issuer['SECRET_KEY']
+                token = generate_jwt_token(self.payload, signing_key)
+                self.assertEqual(utils.jwt_decode_handler(token), self.payload)
+
+    @override_settings(JWT_AUTH=JWT_AUTH_VALID_ISSUERS)
+    def test_decode_raises_invalid_token_error_no_issuers_matched(self):
+        """
+        Validates that a valid token is unable to be decoded given a list of multiple invalid issuers
```

Comment:
Verify the function logs decode failures, and raises `InvalidTokenError` if the given token cannot be decoded.